torchvision/prototype/datasets/_builtin/food101.py

```diff
@@ -0,0 +1,86 @@
+import functools
+from pathlib import Path
+from typing import Any, Tuple, List, Dict, cast, Optional, BinaryIO
+
+from torchdata.datapipes.iter import IterDataPipe, Filter, Mapper, CSVParser, JsonParser, Demultiplexer
+from torchvision.prototype.datasets.utils import Dataset, DatasetInfo, DatasetConfig, HttpResource, OnlineResource
+from torchvision.prototype.datasets.utils._internal import (
+    hint_shuffling,
+    hint_sharding,
+    path_comparator,
+    getitem,
+    INFINITE_BUFFER_SIZE,
+)
+from torchvision.prototype.features import Label, EncodedImage
+
+
+class Food101(Dataset):
+    def _make_info(self) -> DatasetInfo:
+        return DatasetInfo(
+            "food101",
+            homepage="https://data.vision.ee.ethz.ch/cvl/datasets_extra/food-101",
+            # Only train and test splits, no validation.
+            valid_options=dict(split=("train", "test")),
+        )
+
+    def resources(self, config: DatasetConfig) -> List[OnlineResource]:
+        return [
+            HttpResource(
+                url="http://data.vision.ee.ethz.ch/cvl/food-101.tar.gz",
+                sha256="d97d15e438b7f4498f96086a4f7e2fa42a32f2712e87d3295441b2b6314053a4",
+                decompress=True,
+            )
+        ]
+
+    def _classify_archive(self, data: Tuple[str, Any]) -> Optional[int]:
+        path = Path(data[0])
+        if path.parents[1].name == "images":
+            return 0
+        elif path.parents[0].name == "meta":
+            return 1
+        else:
+            return None
+
+    def _prepare_sample(self, data: Tuple[str, BinaryIO]) -> Dict[str, Any]:
+        image_path, image_buffer = data
+        category = Path(image_path).parent.name
+        return dict(
+            image=EncodedImage.from_file(image_buffer),
+            label=Label.from_category(category, categories=self.categories),
+        )
+
+    def _img_in_metadata(self, data: Tuple[str, Any], *, metadata: Dict[str, List[str]]) -> bool:
+        # Check whether or not the file is contained in the list of the splitted metdata.
+        path = Path(data[0])
+        category = Path(path).parent.name
+        if category in metadata:
+            img_ids = {e.split("/")[-1] for e in metadata[category]}
+            return path.stem in img_ids
+        else:
+            return False
```

Comment:
This helper is replaced by a new one for the suggestion I made above.

```suggestion
    def _image_key(self, data: Tuple[str, Any]) -> str:
        path = Path(data[0])
        return path.relative_to(path.parents[1]).with_suffix("").as_posix()
```

Reply:
That's more clean, thanks.